### ISSUE TITLE
[security] Each client should be served from the separate (forked) process with setuid to this user

### ISSUE THREAD

**Issue by yurivict** [paradoxxxzero/butterfly#136]
Currently one root process serves all clients.
This is insecure because in case of the discovery of vulnerabilities client can escalate the privilege level to root. setuid delegates this problem to OS (kernel), which is a better way of dealing with this problem.

```
# pstree  | grep butterfly
 |-+= 65597 root daemon: /usr/local/bin/butterfly.server.py[65598] (daemon)
 | \-+- 65598 root /usr/local/bin/python2.7 /usr/local/bin/butterfly.server.py --login=True --unsecure
```

**Comment by PeterCxy** [comment]
Then we will have to fire up multiple `websocket` servers... which is impractical if the server stands behind a reverse proxy.

**Comment by yurivict** [comment]
Could you please elaborate how does a reverse proxy factor into this?

**Comment by PeterCxy** [comment]
Sorry I misunderstood your issue.  However, currently Butterfly will fire up separate `su` processes for each client session. Do you mean Butterfly should fork itself instead of simply firing up `su` processes?

**Comment by yurivict** [comment]
Yes, I mean butterfly should fork itself and setuid. This will make it marginally more secure, will guard against potential JS or python exploits allowing attackers gaining root access.

**Comment by paradoxxxzero** [comment]
I understand running a web server as root is suboptimal security wise. Due
to the pty.fork the shell session drops the root privileges and gets the
logged user uid/gid etc.. but indeed the parent keeps the root uid but I
don't think os.forking here is a good solution, it would only impact the
websocket to shell stream and the whole web server will still run as root.
(And I'm not sure if tornado will be very happy about this). Or maybe I
misunderstand what you are saying. A perfect solution would be running
everything as a user and delegate the privileges escalation to another
process but it would mean no certificate login as far as I understand.

**Comment by yurivict** [comment]
* When TCP connection is accepted, butterfly python process forks for it.
* When user is logged in, this process setuids.
